pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java

```diff
@@ -0,0 +1,162 @@
+          + " This allows us to estimate table size before onboarding.")
+  public String estimateHeapUsage(String tableSchemaConfigStr,
+      @ApiParam(value = "cardinality in string format", required = true) @QueryParam("cardinality") String cardinality,
+      @ApiParam(value = "primaryKeySize in string format") @QueryParam("primaryKeySize") String primaryKeySize,
+      @ApiParam(value = "numPartitions in string format") @QueryParam("numPartitions") String numPartitionsStr) {
+    ObjectNode resultData = JsonUtils.newObjectNode();
+    TableAndSchemaConfig tableSchemaConfig;
+
+    try {
+      tableSchemaConfig = JsonUtils.stringToObject(tableSchemaConfigStr, TableAndSchemaConfig.class);
+    } catch (IOException e) {
+      throw new ControllerApplicationException(LOGGER,
+          String.format("Invalid TableSchemaConfigs json string: %s", tableSchemaConfigStr),
+          Response.Status.BAD_REQUEST, e);
+    }
+
+    TableConfig tableConfig = tableSchemaConfig.getTableConfig();
+    resultData.put("tableName", tableConfig.getTableName());
+
+    Schema schema = tableSchemaConfig.getSchema();
+
+    // Estimated key space, it contains primary key columns
+    int bytesPerKey = 0;
+    List<String> primaryKeys = schema.getPrimaryKeyColumns();
+
+    if (primaryKeySize != null) {
+      bytesPerKey += Integer.valueOf(primaryKeySize);
+    } else {
+      for (String primaryKey : primaryKeys) {
+        FieldSpec.DataType dt = schema.getFieldSpecFor(primaryKey).getDataType();
+        if (dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST || dt == FieldSpec.DataType.MAP) {
+          String msg = "Not support data types for primary key columns";
+          throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+        } else if (dt == FieldSpec.DataType.STRING) {
+          String msg = "Missing primary key sizes for String columns";
+          throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+        } else {
+          bytesPerKey += dt.size();
+        }
+      }
+      // Java has a 24 bytes array overhead and there's also 8 bytes for the actual array object
+      bytesPerKey += 32;
+    }
+
+    // Estimated value space, it contains <segmentName, DocId, ComparisonValue(timestamp)> and overhead.
+    int bytesPerValue = 64;
+    String comparisonColumn = tableConfig.getUpsertConfig().getComparisonColumn();
+    if (comparisonColumn != null) {
+      FieldSpec.DataType dt = schema.getFieldSpecFor(comparisonColumn).getDataType();
+      if (dt == FieldSpec.DataType.STRING || dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST
+          || dt == FieldSpec.DataType.MAP) {
+        String msg = "Not support data types for the comparison column";
+        throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+      } else {
+        bytesPerValue = 52 + dt.size();
```

Comment:
Is this correct? When comparison column is long, this does not match the timestamp one of 64 bytes

Reply:
thanks @Jackie-Jiang 

For values does it contains offset?
According to discussion in https://github.com/apache/pinot/issues/6729, we need to store RecordLocation, it contains <segmentName, DocId, Timestamp>, so maybe another <48+ 4 + 8>=64 bytes on it.

Reply:
`48 + 4 + 8 = 60 != 64` lol

Reply:
gotcha, thanks. i will update the valueSpace with default timestamp from 64 bytes to 60 bytes.